### ISSUE TITLE
DB Synchronization causes block crash when delegates are not empty

### ISSUE THREAD

**Issue by zachhildreth** [X-CASH-official/XCASH_DPOPS#7]
## Bug for XCASH_DPOPS

This bug has to do with the synchronization process. When the DB synchronization process runs, Sometimes a crash will occur after about 3-4 rounds of block producing. Sometimes it does not crash, but will mess up the memory and loose some of the variables and not be able to vote on the block. It will do one of these things most of the times

**When the DB sync is turned off this problem does not occur, and the block producing process can run for hours. **

A debugger will point you to a DB function, but I am pretty sure its not the problem as the debugger will sometimes point you to a different function, and this general means the memory gets messed up and that none of those functions are the real cause. (read the facts for more info)

Also all db functions are run in the test, and run 1000 times to be checked for memory leaks and they never create a problem

How to turn off DB sync at bootup
You can do this by going to `src/XCASH_DPOPS.c` and removing this code:
```
// check if the database is synced, unless this is the main network data node
  if (memcmp(xcash_wallet_public_address,NETWORK_DATA_NODE_1_PUBLIC_ADDRESS,XCASH_WALLET_LENGTH) != 0)
  {
    // check if all of the databases are synced from a network data node, since their is no way to tell if the previous round could not reach consensus
    if (check_if_databases_are_synced(2) == 0)
    {
      MAIN_ERROR("Could not check if the databases are synced");
    }
  }
```

If you remove the above code, the block producing process will run at 100% validation for hours. If you run the above code, the block producing process will most likely fail on one of the delegates within 5 blocks, but will not fail on all of them.

## Some facts about this bug
* Tried upping all stack allocated variables length
* Tried upping the delegates initialize struct amount in `src/global_data/initialize_and_reset_structs_define_macros`
* Checked to make sure all string lengths were correct
* Checked for any calloc or malloc that did not have a free
* Checked any mistakes of using sizeof to clear a heap allocated variable

### Result of this bug
It causes the whole program to segfault  and crash after a few rounds, or gets it off the majority path of block producing, and it will sit out that round and sometimes work the next round and sometimes not due to the memory mess up

### Status of bug
I will keep looking into this bug but have spent a few days on it, worst case is I will make a double boot so it will sync on reboot and then reboot again to run the regular program.

We will welcome any Pull request or comments that follows the [XCASH_DPOPS bug bounty guidelines](https://docs.xcash.foundation/contribute/bug-bounty-program)

### We have determined that the [CVSS score for this bug](https://nvd.nist.gov/vuln-metrics/cvss/v3-calculator?vector=AV:L/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H/E:X/RL:X/RC:R/CR:X/IR:X/AR:H/MAV:L/MAC:L/MPR:N/MUI:N/MS:U/MC:X/MI:X/MA:H&version=3.1) is a 7.7

So according to the [Bug Bounty official documentation](https://docs.xcash.foundation/contribute/bug-bounty-program) this bug qualifies for a award between $250.00 and $500.00 USD

We have calculated the final amount in USD based on the following criteria

| Question                                     | Answer  |
| -------------------------------------------- | --------------------------- |
| How long I have spent on the bug                                      | about 30 hours  |
| How often does the bug occur                                            | depends if the sync is run |
| Was there any modification to the code to create the bug  | NO |
| Did the user provide a full code fix                                | ? |
| Did the bug have to do with multi-threading                       | ? |
| Could the tester use a debugger, static analysis or both |both|
| Are we currently in Alpha, Beta or Live implementation of DPOPS |ALPHA|

 

## Based on the above it will be around 350-400 depending on what happens to the ? Questions

**Comment by xihuan22d** [comment]
in submit_block_template function，use memset(data,0,strlen(data));   
 //set the data ,but send_http_request  use the date ,data is not thread safe，send_http_request  is asynchronous ,so in the  

for (count = 0; count < NETWORK_DATA_NODES_AMOUNT; count++) ,it  will Multi threads


int submit_block_template(char* data, const int HTTP_SETTINGS)
{
  // Constants
  const char* HTTP_HEADERS[] = {"Content-Type: application/json","Accept: application/json"}; 
  const size_t HTTP_HEADERS_LENGTH = sizeof(HTTP_HEADERS)/sizeof(HTTP_HEADERS[0]);
  const size_t DATA_LENGTH = strnlen(data,BUFFER_SIZE);

  // Variables
  char message[BUFFER_SIZE];

  // define macros
  #define SUBMIT_BLOCK_TEMPLATE_ERROR(settings) \
  memcpy(error_message.function[error_message.total],"submit_block_template",21); \
  memcpy(error_message.data[error_message.total],settings,sizeof(settings)-1); \
  error_message.total++; \
  return 0;
  
  memset(message,0,sizeof(message));

  // create the message
  memcpy(message,"{\"jsonrpc\":\"2.0\",\"id\":\"0\",\"method\":\"submit_block\",\"params\":[\"",61);
  memcpy(message+61,data,DATA_LENGTH);
  memcpy(message+61+DATA_LENGTH,"\"]}",3);
  memset(data,0,strlen(data));    //set the data ,but send_http_request  use the date  ,data is not thread safe

  if (send_http_request(data,"127.0.0.1","/json_rpc",XCASH_DAEMON_PORT,"POST", HTTP_HEADERS, HTTP_HEADERS_LENGTH,message,RECEIVE_DATA_TIMEOUT_SETTINGS,"submit block template",HTTP_SETTINGS) <= 0)
  {  
    SUBMIT_BLOCK_TEMPLATE_ERROR("Could not submit the block template");
  }
  
  if (strstr(data,"error") != NULL)
  {
    SUBMIT_BLOCK_TEMPLATE_ERROR("Could not submit the block template");
  }
  return 1;

  #undef SUBMIT_BLOCK_TEMPLATE_ERROR
}

**Comment by xihuan22d** [comment]
I don't find send_http_request contain any thread code，I naturally think that it is asynchronous and ignore  it .
But my instinct tells me that this problem should be related to thread safe . I will continue to observe these related methods.

**Comment by zachhildreth** [comment]
The bug has been fixed in commit [504a9f461940072fc85fd1afc1836df9b5e64878](https://github.com/X-CASH-official/XCASH_DPOPS/commit/504a9f461940072fc85fd1afc1836df9b5e64878)

The error message given on one of the delegates was always "could not get the previous block hash" This in turn would stop the block creation since the test are always run at 100% block validity instead of the 67% needed for a successful DBFT round.

The reason for the crash was the way the sign and more importantly verify message function was using the RPC calls. By design it was going to use over 3000 RPC calls in 50 seconds, and was only using about 100 in 50 seconds to overload the RPC wallet since the xcash-wallet-rpc is not multi threaded

The solution was to redesign how the messages between delegates will be signed, so instead of using the private key of the wallet, it will now be using a separate ECDSA key using the VRF code to sign and verify messages.

This change not only dropped the RPC usage to just a few calls per block now, but has also made the DB syncing process about about ~3-5X faster and the message throughput for the epoll server about ~3-5X more depending on the internet connection and speed of the computer.
